packages/ra-core/src/dataProvider/useDataProviderWithDeclarativeSideEffects.ts

```diff
@@ -0,0 +1,73 @@
+import {
+    useNotify,
+    useRedirect,
+    useRefresh,
+    useUnselectAll,
+} from '../sideEffect';
+import useDataProvider, { DataProviderHookFunction } from './useDataProvider';
+import { useCallback } from 'react';
+
+const useDataProviderWithDeclarativeSideEffects = (): DataProviderHookFunction => {
+    const dataProvider = useDataProvider();
+    const notify = useNotify();
+    const redirect = useRedirect();
+    const refresh = useRefresh();
+    const unselectAll = useUnselectAll();
+
+    return useCallback(
+        (type: string, resource: string, params: any, options: any = {}) => {
+            const convertToFunctionSideEffect = sideEffects => {
+                if (
+                    typeof sideEffects === 'undefined' ||
+                    typeof sideEffects === 'function'
+                ) {
+                    return sideEffects;
```

Comment:
You can test `null` too. But it's not possible using `typeof`, because `null` has type 'object'`.

Reply:
Reviewed